src/core/public/core_app/errors/error_application.tsx

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { ReactChild, useState, useLayoutEffect } from 'react';
+import ReactDOM from 'react-dom';
+import { History } from 'history';
+import { i18n } from '@kbn/i18n';
+import { I18nProvider } from '@kbn/i18n/react';
+
+import { EuiEmptyPrompt, EuiPage, EuiPageBody, EuiPageContent } from '@elastic/eui';
+import { UrlOverflowUi } from './url_overflow_ui';
+import { IBasePath } from '../../http';
+import { AppMountParameters } from '../../application';
+
+interface Props {
+  title?: string;
+  children?: ReactChild;
+}
+
+const AppError: React.FC<Props> = ({ title, children }) => {
+  title =
+    title ??
+    i18n.translate('core.application.appRenderError.defaultTitle', {
+      defaultMessage: 'Application error',
+    });
+
+  return (
+    <EuiPage style={{ minHeight: '100%' }} data-test-subj="appRenderErrorPageContent">
+      <EuiPageBody>
+        <EuiPageContent verticalPosition="center" horizontalPosition="center">
+          <EuiEmptyPrompt
+            iconType="alert"
+            iconColor="danger"
+            title={<h2>{title}</h2>}
+            body={children}
+          />
+        </EuiPageContent>
+      </EuiPageBody>
+    </EuiPage>
+  );
+};
+
+const ErrorApp: React.FC<{ basePath: IBasePath; history: History }> = ({ basePath, history }) => {
+  const [currentLocation, setCurrentLocation] = useState(history.location);
+  useLayoutEffect(() => history.listen((location) => setCurrentLocation(location)), [history]);
```

Comment:
> You should return an unregister/cleanup handler from useLayoutEffect to avoid possibly having multiple listeners hooked to the history.

`history.listen` returns a unregister callback, but I can see how it's not obvious looking at this code. I'll go ahead and `{ return ... }` to make it more clear.

> Also why `useLayoutEffect` instead of a plain `useEffect` here? It doesn't seems like it's DOM related?

React hook execution order is hard to understand, but essentially what I found in testing is that `useEffect` was executing after the component was mounted and after the first history update was dispatched. `useLayoutEffect` does not have this problem since it fires immediately after the component is mounted, thus not losing any history updates. (Though I suspect it could miss an update in Concurrent mode, depending on the implementation?). The docs explain this:

> useEffect fires after layout and paint, during a deferred event

> However, not all effects can be deferred...For these types of effects, React provides one additional Hook called useLayoutEffect.

> The signature is identical to useEffect, but it fires synchronously after all DOM mutations. Updates scheduled inside useLayoutEffect will be flushed synchronously, before the browser has a chance to paint.

IMO the name of `useLayoutEffect` is too specific to the expected usage of this API.

Reply:
> history.listen returns a unregister callback, but I can see how it's not obvious looking at this code

Oh right, my bad. 

> React hook execution order is hard for me to understand, but essentially what I found in testing is that useEffect was executing after the component was mounted and after the first history update was dispatched

Yea, took me a while to understand the exact difference between these two. They are fired 'virtually' at the same time, after each component update, the difference being that `useEffect` is triggered asynchronously (setImmediate or nextTick I don't remember), where `useLayoutEffect` is synchronous, therefor effectively executed before repaint.

> thus not losing any history updates

it's fine keeping `useLayoutEffect` anyway, but just to try to understand:

I'm surprised you had an issue using `useEffect`, as the current location is initialized in your `useState` hook using `const [currentLocation, setCurrentLocation] = useState(history.location);`. Only missable location change would therefor be between the initial mounting, and the end of the initial rendering. The issues you encountered where in `enzyme`-based tests, right?

Reply:
TBH I'm encountering a lot of trouble when testing FC with effects using enzyme, even when the library is supposed to now be 100% working with FCs...